### PR TITLE
Reinitialize geometric search when mesh changes and test 

### LINKED ISSUE #6799: Geometric search not reinitialized when mesh changes
### Description of the enhancement or error report

When the mesh changes due to adaptivity or xfem, the geometric search data is updated rather than reinitialized, so newly-created nodes on the slave surface are not included in the contact search.
### Rationale for the enhancement or information for reproducing the error

This is needed for using adaptvity or xfem in conjunction with contact.
### Identified impact

Minor change to DisplacedProblem.C

### PR DESCRIPTION
**Pull request by bwspenc**
closes #6799

### CODE REVIEW

**Comment by bwspenc** [comment]
contact_adaptivity_fix
Before (left) and After (right)

**Comment by jwpeterson** [comment]
Nice, I think you may have discovered some new physics there on the left.  "ghost clad"

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -599,8 +599,8 @@ MooseMesh::nodeToElemMap()
     Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
     if (!_node_to_elem_map_built)
     {
-      MeshBase::const_element_iterator       el  = getMesh().elements_begin();
-      const MeshBase::const_element_iterator end = getMesh().elements_end();
+      MeshBase::const_element_iterator       el  = getMesh().active_elements_begin();
+      const MeshBase::const_element_iterator end = getMesh().active_elements_end();
```
Comment (bwspenc) [inline framework/src/mesh/MooseMesh.C]: Without this change, we hit an assert in NonlinearSystem::findImplicitGeometricCouplingEntries().  I think this is the right thing to do.  I would expect that nodeToElemMap() would only return active elements.

**Comment by bwspenc** [comment]
It looks like I stepped into something that was recently hashed out in #6032 and #6040 (pinging @dschwen and @friedmud).  My problem is that the equation solver reinit gets solved before MooseMesh::meshChanged() gets called. Within the equation solver reinit, NonlinearSystem::findImplicitGeometricCouplingEntries() gets called, and for that we need only the active elements in _node_to_elem_map.

I tried using _node_to_active_semilocal_elem_map instead, but still hit an assert. We probably don't just want the semilocal elements either.

The real problem is that for _node_to_elem_map to be correct in the equation solver reinit, it has to be updated, but that happens right after that.  I tried switching the order of the calls to _eq.reinit() and _mesh.meshChanged() in FEProblem and DisplacedProblem, and that fixes my issue, but then I hit asserts in other tests that use adaptivity.

**Comment by bwspenc** [comment]
The test is failing in debug because of libmesh issue [#938](https://github.com/libMesh/libmesh/issues/938).

**Comment by tonkmr** [comment]
I could really use these changes for some of the work one of my students is doing. Is any progress being made on libmesh issue [#938](https://github.com/libMesh/libmesh/issues/938)?

**Comment by jwpeterson** [comment]
I started to look at it back before the long weekend, but got sidetracked.  @bwspenc, what is the easiest way for me to reproduce this again?  Did you say you had a libmesh-only test case for that?

**Comment by bwspenc** [comment]
@jwpeterson, the regression test that I checked in with this branch is probably the best test case. I didn't have a libmesh-only test case. It's been a little while, but I think the path forward is to figure out a way to skip `contract()`, and then resolve the issue we were having when we tried to create exodus output after doing that.

**Comment by jwpeterson** [comment]
OK, I will take a look.

**Comment by jwpeterson** [comment]
The bug in libmesh, which was related to writing Exodus solution vectors with NULL entries in them, was fixed a while ago.  This PR now produces a different error message which @bwspenc is working on.

**Comment by jwpeterson** [comment]
Closing this for now as the original issue in libmesh was fixed a while ago, and there does not seem to be any development currently going on.

**Comment by bwspenc** [comment]
@roystgnr This is the branch.

**Comment by roystgnr** [comment]
The DisplacedProblem.C change doesn't regress any displaced/displacement tests for me, even with DistributedMesh and recovery turned on, so that's a plus.  I'll check out the new modules test next.

**Comment by roystgnr** [comment]
Slight update for deprecated syntax in roystgnr/roys_geomsearch_reinit

On the new test I'm getting a libMesh assertion falure - NonlinearSystemBase::findImplicitGeometricCouplingEntries is asking for dof_indices on an inactive element.  I'll see if I can fix it.

**Comment by bwspenc** [comment]
@roystgnr That sounds like the problem I was hitting back when I last looked at this. It's been a while since I looked at it, but I think the problem stemmed from the fact that the two meshes were briefly out of sync (one had the elements removed and the other didn't) when that geometric search happened. It was pretty messy!

**Pull request closed without merging**